receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+var (
+	errScrape = errors.New("failed to successfully scrape any SNMP metrics")
+)
+
+var (
+	errMsgBadValueType                  = `returned metric SNMP data type for OID: %s is not supported`
+	errMsgIndexedBadValueType           = `returned metric SNMP data type for OID: %s from column OID: %s is not supported`
+	errMsgBadIndexedAttributes          = `problem retrieving SNMP indexed attribute data: %w`
+	errMsgBadIndexedResourceAttributes  = `problem retrieving SNMP indexed resource attribute data: %w`
+	errMsgIndexedAttributesBadValueType = `returned attribute SNMP data type for OID: %s from column OID: %s is not supported`
+)
+
+// snmpScraper handles scraping of SNMP metrics
+type snmpScraper struct {
+	client   client
+	logger   *zap.Logger
+	cfg      *Config
+	settings component.ReceiverCreateSettings
+}
+
+// newScraper creates an initialized snmpScraper
+func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCreateSettings) *snmpScraper {
+	return &snmpScraper{
+		logger:   logger,
+		cfg:      cfg,
+		settings: settings,
+	}
+}
+
+// start gets the client ready
+func (s *snmpScraper) start(ctx context.Context, host component.Host) (err error) {
+	s.client, err = newClient(s.cfg, host, s.settings.TelemetrySettings, s.logger)
+	if err != nil {
+		return err
+	}
+	err = s.client.Connect()
+
+	return
+}
+
+// shutdown closes the client
+func (s *snmpScraper) shutdown(ctx context.Context, host component.Host) (err error) {
+	err = s.client.Close()
+
+	return
+}
+
+// scrape collects and creates OTEL metrics from a SNMP environment
+func (s *snmpScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	// Get a basic ResourceMetrics prepped for metrics with no resource attributes
+	resourceMetrics := pmetric.NewResourceMetrics()
+	scopeMetrics := resourceMetrics.ScopeMetrics().AppendEmpty()
+	scopeMetrics.Scope().SetName("otelcol/snmpreceiver")
+	scopeMetrics.Scope().SetVersion(s.settings.BuildInfo.Version)
+	metricSlice := scopeMetrics.Metrics()
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	metricsProcessed := false
+
+	resourceMetricsMap := map[string]*pmetric.ResourceMetrics{}
+	metricsMap := map[string]*pmetric.Metric{}
+	// Try to scrape scalar OID based metrics.
+	// metricsMap is passed in as a place for any created metrics to live
+	if err := s.scrapeScalarMetrics(now, metricsMap); err != nil {
+		s.logger.Warn("Failed to collect scalar OID metrics", zap.Error(err))
+	} else if len(metricsMap) != 0 {
+		metricsProcessed = true
+		// Add new scalar metrics to the basic ResourceMetrics with no resource attributes
+		for _, value := range metricsMap {
+			value.MoveTo(metricSlice.AppendEmpty())
+		}
+		// Load the map with the basic ResourceMetrics (that possibly
+		// has scalar OID based metrics already on it)
+		resourceMetricsMap[""] = &resourceMetrics
```

Comment:
Is the empty string represent the scalar metrics? If so it might be more straight forward to create a const.

```go
const scalerMapKey = ""

...
resourceMetricsMap[scalerMapKey] = &resourceMetrics
```

Reply:
not a bad idea. i changed this where applicable